libraries/AP_Notify/NeoPixel_Oreo.cpp

```diff
@@ -0,0 +1,113 @@
+/*
+   This program is free software: you can redistribute it and/or modify
+   it under the terms of the GNU General Public License as published by
+   the Free Software Foundation, either version 3 of the License, or
+   (at your option) any later version.
+
+   This program is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+   GNU General Public License for more details.
+
+   You should have received a copy of the GNU General Public License
+   along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "NeoPixel_Oreo.h"
+#include "NeoPixel.h"
+
+extern const AP_HAL::HAL& hal;
+
+NeoPixel_Oreo::NeoPixel_Oreo(uint8_t theme) :
+    NotifyDevice()
+{
+    oreo = new OreoLED_I2C(0,theme);
+}
+
+//
+// Initialize the LEDs
+//
+bool NeoPixel_Oreo::init()
+{
+    enable_mask = NeoPixel::init_ports();
+    hal.scheduler->register_timer_process(FUNCTOR_BIND_MEMBER(&NeoPixel_Oreo::timer, void));
+    return true;
+}
+
+void NeoPixel_Oreo::update()
+{
+    oreo->update();
+}
+
+void NeoPixel_Oreo::timer()
+{
+    WITH_SEMAPHORE(_sem);
+
+    const uint32_t now_ms = AP_HAL::millis();
+    if (now_ms - _last_init_ms >= 1000) {
+        _last_init_ms = now_ms;
+        enable_mask = NeoPixel::init_ports();
+    }
+    if (enable_mask == 0) {
+        // nothing is enabled, no pins set as LED output
+        return;
+    }
+
+    // for each LED
+    for (uint8_t i=0; i<OREOLED_NUM_LEDS; i++) {
+        // check for state change
+        switch (oreo->_state_desired[i].mode) {
+        case OreoLED_I2C::OREOLED_MODE_MACRO:
+            if (now_ms - oreo->_state_desired[i].last_update_ms >= 2000) {
```

Comment:
better to put phase offset here, on right side as adjustment to the 2000, and not change last_update_ms

Reply:
fixed